libminifi/src/core/ProcessSession.cpp

```diff
@@ -241,6 +241,18 @@ void ProcessSession::write(const std::shared_ptr<core::FlowFile> &flow, OutputSt
   }
 }
 
+void ProcessSession::writeBuffer(const std::shared_ptr<core::FlowFile>& flow_file, gsl::span<const char> buffer) {
+  struct BufferOutputStreamCallback : OutputStreamCallback {
```

Comment:
I'd rather we move out this `SpanOutputStreamCallback`, create a forwarding `ProcessSession::write` that takes a reference so we don't have to create local callback variables, and call that

Reply:
I don't fully understand what you mean here.
What would this forwarding `write` take?
What's the problem with local callback variables? Lambdas are ruled out because of the `OutputStreamCallback` interface, but a preference for local scope remains.

Reply:
What do you think about a `write` that takes a lambda and a `writeBuffer` that passes a lambda to this new `write`?

Reply:
I recall a refactor concerning taking addresses of non-lvalues, due to this we have some (not horribly many) instances of creating local callbacks for the sole purpose of taking their addresses, creating a `ProcessSession(const shrd_ptr<FF>&, OutputStreamCallback&&)` would handle these "forwarding" use-cases (where the temporary callbacks only forward the stream data in some implementation dependent manner),
I just would like to avoid the proliferation of "input/output-dependent" methods on `ProcessSession` (like `writeFile`, `writeRawBuffer`, `writeString`, etc.)

Reply:
> I just would like to avoid the proliferation of "input/output-dependent" methods on ProcessSession (like writeFileStream, writeRawBuffer, writeString, etc.)

I don't want this either, but a buffer is a very basic concept and it's expected by almost all write functions in existence. A `gsl::span` is just a pointer and a size. I believe `writeBuffer` is the most basic and generic interface possible to write to a flow file, and therefore it should be in `ProcessSession`.

> I recall a refactor concerning taking addresses of non-lvalues, due to this we have some (not horribly many) instances of creating local callbacks for the sole purpose of taking their addresses, creating a ProcessSession(const shrd_ptr<FF>&, OutputStreamCallback&&) would handle these "forwarding" use-cases (where the temporary callbacks only forward the stream data in some implementation dependent manner),

What's your opinion about this? https://pastebin.com/q7d67x1j

Reply:
compared to the current implementation, I would go with the current one, agree that a buffer is quite fundamental, one other idea we could entertain is a `write(const shrd_ptr<FF>&, const Container&)` template, but we could generalize it later when there is actual need for it, for now `writeBuffer` seems to do the job